source/grid/grid_tools.cc

```diff
@@ -1637,6 +1635,187 @@ namespace GridTools
+  {
+    std::vector<typename MeshType::active_cell_iterator> subdomain_boundary_cells, active_cell_layer_within_distance;
+    std::vector<bool> vertices_outside_subdomain ( mesh.get_triangulation().n_vertices(),
+                                                   false);
+
+    const unsigned int spacedim = MeshType::space_dimension;
+
+    unsigned int n_non_predicate_cells = 0; // Number of non predicate cells
+
+    // Find the layer of cells for which predicate is true and that
+    // are on the boundary with other cells. These are
+    // subdomain boundary cells.
+
+    // Find the cells for which the predicate is false
+    // These are the cells which are around the predicate subdomain
+    for ( typename MeshType::active_cell_iterator
+          cell = mesh.begin_active();
+          cell != mesh.end(); ++cell)
+      if ( !predicate(cell)) // Negation of predicate --> Not Part of subdomain
+        {
+          for (unsigned int v=0; v<GeometryInfo<MeshType::dimension>::vertices_per_cell; ++v)
+            vertices_outside_subdomain[cell->vertex_index(v)] = true;
+          n_non_predicate_cells++;
+        }
+
+    // If all the active cells confirm to the predicate
+    // or if none of the active cells confirm to the predicate
+    // there is no active cell layer around the predicate
+    // subdomain (within any distance)
+    if ( n_non_predicate_cells == 0  || n_non_predicate_cells == mesh.get_triangulation().n_active_cells() )
+      return std::vector<typename MeshType::active_cell_iterator>();
+
+    // Find the cells that conform to the predicate
+    // but share a vertex with the cell not in the predicate subdomain
+    for ( typename MeshType::active_cell_iterator
+          cell = mesh.begin_active();
+          cell != mesh.end(); ++cell)
+      if ( predicate(cell)) // True predicate --> Potential boundary cell of the subdomain
+        for (unsigned int v=0; v<GeometryInfo<MeshType::dimension>::vertices_per_cell; ++v)
+          if (vertices_outside_subdomain[cell->vertex_index(v)] == true)
+            {
+              subdomain_boundary_cells.push_back(cell);
+              break; // No need to go through remaining vertices
+            }
+
+    // To cheaply filter out some cells located far away from the predicate subdomain,
+    // get the bounding box of the predicate subdomain.
+    std::pair< Point<spacedim>, Point<spacedim> > bounding_box = compute_bounding_box( mesh,
+        predicate );
+    // Add layer_thickness to the bounding box
+    for ( int d=0; d<spacedim; ++d)
+      {
+        bounding_box.first[d]  -= layer_thickness; // minp
+        bounding_box.second[d] += layer_thickness; // maxp
+      }
+
+    std::vector< Point<spacedim> > boundary_cells_centers; // cache all the subdomain boundary cells centers here
+    std::vector< double> boundary_cells_radii; // cache all the subdomain boundary cells radii
```

Comment:
Also, these two variables correspond to the `subdomain_boundary_cells`. It would be useful to name them `subdomain_boundary_cells_centers` and `subdomain_boundary_cells_radii` -- that's a long name, but it would have avoided what I just asked myself, namely "what boundary".

You can make the following operations more efficient if you insert
```
  subdomain_boundary_cells_centers.reserve (subdomain_boundary_cells.size();
  subdomain_boundary_cells_radii.reserve (subdomain_boundary_cells.size();
```